core/embed/trezorhal/stm32f4/syscall_dispatch.c

```diff
@@ -105,318 +121,419 @@ __attribute((no_stack_protector)) void syscall_handler(uint32_t *args,
     } break;
+
     case SYSCALL_SDCARD_IS_PRESENT: {
       args[0] = sdcard_is_present();
     } break;
+
     case SYSCALL_SDCARD_GET_CAPACITY: {
       args[0] = sdcard_get_capacity_in_bytes();
     } break;
+
     case SYSCALL_SDCARD_READ_BLOCKS: {
-      args[0] = sdcard_read_blocks((uint32_t *)args[0], args[1], args[2]);
+      uint32_t *dest = (uint32_t *)args[0];
+      uint32_t block_num = args[1];
+      uint32_t num_blocks = args[2];
+      args[0] = sdcard_read_blocks__verified(dest, block_num, num_blocks);
     } break;
+
     case SYSCALL_SDCARD_WRITE_BLOCKS: {
-      args[0] =
-          sdcard_write_blocks((const uint32_t *)args[0], args[1], args[2]);
+      const uint32_t *src = (const uint32_t *)args[0];
+      uint32_t block_num = args[1];
+      uint32_t num_blocks = args[2];
+      args[0] = sdcard_write_blocks__verified(src, block_num, num_blocks);
     } break;
 #endif
 
     case SYSCALL_UNIT_VARIANT_PRESENT: {
       args[0] = unit_variant_present();
     } break;
+
     case SYSCALL_UNIT_VARIANT_GET_COLOR: {
       args[0] = unit_variant_get_color();
     } break;
+
     case SYSCALL_UNIT_VARIANT_GET_PACKAGING: {
       args[0] = unit_variant_get_packaging();
     } break;
+
     case SYSCALL_UNIT_VARIANT_GET_BTCONLY: {
       args[0] = unit_variant_get_btconly();
     } break;
+
     case SYSCALL_UNIT_VARIANT_IS_SD_HOTSWAP_ENABLED: {
       args[0] = unit_variant_is_sd_hotswap_enabled();
     } break;
+
     case SYSCALL_SECRET_BOOTLOADER_LOCKED: {
       args[0] = secret_bootloader_locked();
     } break;
+
 #ifdef USE_BUTTON
     case SYSCALL_BUTTON_READ: {
       args[0] = button_read();
     } break;
+
     case SYSCALL_BUTTON_STATE_LEFT: {
       args[0] = button_state_left();
     } break;
+
     case SYSCALL_BUTTON_STATE_RIGHT: {
       args[0] = button_state_right();
     } break;
 #endif
+
 #ifdef USE_TOUCH
     case SYSCALL_TOUCH_GET_EVENT: {
       args[0] = touch_get_event();
     } break;
 #endif
+
 #ifdef USE_HAPTIC
     case SYSCALL_HAPTIC_SET_ENABLED: {
-      haptic_set_enabled(args[0]);
+      bool enabled = (args[0] != 0);
+      haptic_set_enabled(enabled);
     } break;
+
     case SYSCALL_HAPTIC_GET_ENABLED: {
       args[0] = haptic_get_enabled();
     } break;
+
     case SYSCALL_HAPTIC_TEST: {
-      args[0] = haptic_test(args[0]);
+      uint16_t duration_ms = (uint16_t)args[0];
+      args[0] = haptic_test(duration_ms);
     } break;
+
     case SYSCALL_HAPTIC_PLAY: {
-      args[0] = haptic_play(args[0]);
+      haptic_effect_t effect = (haptic_effect_t)args[0];
+      args[0] = haptic_play(effect);
     } break;
+
     case SYSCALL_HAPTIC_PLAY_CUSTOM: {
-      args[0] = haptic_play_custom(args[0], args[1]);
+      int8_t amplitude_pct = (int8_t)args[1];
+      uint16_t duration_ms = (uint16_t)args[2];
+      args[0] = haptic_play_custom(amplitude_pct, duration_ms);
     } break;
 #endif
 
 #ifdef USE_OPTIGA
-      /*optiga_sign_result optiga_sign(uint8_t index, const uint8_t *digest,
```

Comment:
Shouldn't we have a `SYSCALL_OPTIGA_SIGN` though?

Reply:
Yes, thanks for catching that. I've added fixed it, see 2cff1e570d4fbbf83400e8da0044dd1418a747d1